Modifier to apply energy corrections to photons and electrons (miniAOD) for 74X

matteosan1 opened the pull request:
Same as #10611.
Weight files are temporarily stored in RecoEgamma/EgammaTools/data. They will removed once the weights will be integrated into the GT.
Working on comparison plots for the resolution.

matteosan1 committed:
first commit fot 74X

cmsbuild commented:
A new Pull Request was created by @matteosan1 (Matteo Sani) for CMSSW_7_4_X.

Modifier to apply energy corrections to photons and electrons (miniAOD) for 74X

It involves the following packages:

PhysicsTools/PatAlgos
RecoEgamma/EgammaTools
RecoEgamma/ElectronIdentification
RecoEgamma/PhotonIdentification


@cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please review it and eventually sign? Thanks.
@rappoccio, @Sam-Harper, @imarches, @ahinzmann, @acaudron, @mmarionncern, @lgray, @jdolen, @nhanvtran, @schoef, @ferencek, @gpetruc, @mariadalfonso, @pvmulder, @TaiSakuma this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

lgray commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f9c91941766585755aa15c6f1875423645b932df
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7031/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/ea91159ea28ec861ee2aec302292540e4911553a
https://github.com/cms-sw/cmssw/commit/fbc8fecbf36cc81e0a9279d869e160d5b65d679d
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7031/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7031/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7031/summary.html

matteosan1 commented:
I have updated this PR to read from the DB. The payloads have been loaded into the dropbox.

cmsbuild commented:
Pull request #10613 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: d4405fd151013e0eabaff93811e8771eb771c707
 When I ran the RelVals I found an error in the following worklfows: 
 50202.0 step3
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50/step3_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50.log

----- Begin Fatal Exception 08-Aug-2015 02:10:19 CEST-----------------------

An exception of category 'Configuration' occurred while

   [0] Processing run: 1 lumi: 1 event: 1

   [1] Running path 'validation_step'

   [2] Calling event method for module MixingModule/'mix'

Exception Message:

RootInputFileSequence::readOneSpecified(): no input files specified for secondary input source.

----- End Fatal Exception -------------------------------------------------

</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7080/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/e629502d4b2f7736b95338dbf55d7dbb9abc7b60
https://github.com/cms-sw/cmssw/commit/ea91159ea28ec861ee2aec302292540e4911553a
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7080/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7080/git-merge-result

matteosan1 committed:
removed .root

configure all the value map producers for MiniAOD production

fix to run with data

final updates to read from DB

code simplification

slava77 commented:
@cmsbuild please test
hoping for no DAS error this time

slava77 commented:
please address issues reported by the static analyzer for the code changes in this PR https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7080/llvm-analysis/

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented on RecoEgamma/EgammaPhotonProducers/python/reducedEgamma_cfi.py:
```diff
@@ -58,4 +58,118 @@
 )
 
+from CondCore.DBCommon.CondDBCommon_cfi import CondDBCommon
+gedelectronGBRESSource = cms.ESSource("PoolDBESSource",
```
I'm expecting that this will go away from this PR by the time it's merged.
OK to keep for now while testing.

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -0,0 +1,730 @@
+
+#include "RecoEgamma/EgammaElectronAlgos/interface/RegressionHelper.h"
+
+#include "TFile.h"
```
this is not used, please remove

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -0,0 +1,730 @@
+#include <vdt/vdtMath.h>
+
+namespace {
+  //const edm::EDGetTokenT<edm::ValueMap<float> > empty_token;
```
please drop the commented out line if not needed

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -0,0 +1,730 @@
+  struct electron_config {
+    edm::InputTag electron_src;
+    edm::EDGetTokenT<edm::View<pat::Electron> > tok_electron_src;
+    edm::InputTag sigmaIetaIphi;
```
Maybe define a ValueMapTagTokenPair struct and use it throughout.
Keeping these in sync otherwise appears tedious and error prone

slava77 replied:
if you also put these config element structs into a vector, a lot of the code can be simplified where the values don't need to be spelled out

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -0,0 +1,730 @@
+	  run > 253000) {
+	bunchspacing_ = 25;
+      } else {
+	bunchspacing_ = 50;
```
make this in sync with 
https://github.com/cms-sw/cmssw/blob/CMSSW_7_4_X/RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc#L140
[this temporary hack that we introduced is becoming a norm :( ]

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -0,0 +1,730 @@
+    }
+  }
+
+  std::array<float, 33> eval;
```
the format better have been defined somewhere so that proper alignment of parameters is more reliable to code

matteosan1 replied:
Hi Slava,
I am implementing the changes you requested.
What do you mean with this one  (or better what is the solution you are 
proposing) ?

On 8/8/15 22:53, Slava Krutelyov wrote:
>
> In RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc 
> <https://github.com/cms-sw/cmssw/pull/10613#discussion_r36581750>:
>
> > +  // we encounter two cases here, either we are running AOD -> MINIAOD
> > +  // and the value maps are to the reducedEG object, can use original object ptr
> > +  // or we are running MINIAOD->MINIAOD and we need to fetch the pat objects to reference
> > +  edm::Ptr<reco::Candidate> ptr(ele.originalObjectRef());
> > +  if( !e_conf.tok_electron_src.isUninitialized() ) {
> > +    auto key = eles_by_oop.find(ele.originalObjectRef().key());
> > +    if( key != eles_by_oop.end() ) {
> > +      ptr = key->second;
> > +    } else {
> > +      throw cms::Exception("BadElectronKey")
> > +        << "Original object pointer with key = " << ele.originalObjectRef().key()
> > +        << " not found in cache!";
> > +    }
> > +  }
> > +
> > +  std::array<float, 33> eval;
>
> the format better have been defined somewhere so that proper alignment 
> of parameters is more reliable to code
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cms-sw/cmssw/pull/10613/files#r36581750>.
>

+++++++++++
Matteo Sani
University of California, San Diego, US.
Tel: +41227671577
CERN Office: 40-3-A02

cmsbuild commented:
+1
Tested at: d4405fd151013e0eabaff93811e8771eb771c707
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7085/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/e629502d4b2f7736b95338dbf55d7dbb9abc7b60
https://github.com/cms-sw/cmssw/commit/ea91159ea28ec861ee2aec302292540e4911553a
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7085/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7085/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7085/summary.html

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -0,0 +1,730 @@
+  assignValue(ptr, e_conf.tok_cryPhi, ele_vmaps, cryPhi);
+  assignValue(ptr, e_conf.tok_cryEta, ele_vmaps, cryEta);
+
+  eval[8]  = eMax/sc->rawEnergy();
```
enums maybe?
We were discussing with Lindsey in another PR to have something like TMVAEvaluator for this kind of data manipulation.
I guess a solution there is coming soon.
@lgray

lgray replied:
@slava77 added to nearer-term stack. I'll see what I can do.

On Mon, Aug 10, 2015 at 1:51 PM, Slava Krutelyov <notifications@github.com>
wrote:

> In RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc
> <https://github.com/cms-sw/cmssw/pull/10613#discussion_r36623288>:
>
> > +  assignValue(ptr, e_conf.tok_clusterMaxDRRawEnergy , ele_vmaps, clusterMaxDRRawEnergy);
> > +  assignValue(ptr, e_conf.tok_clusterRawEnergy0	    , ele_vmaps, clusterRawEnergy0    );	
> > +  assignValue(ptr, e_conf.tok_clusterRawEnergy1	    , ele_vmaps, clusterRawEnergy1    );	
> > +  assignValue(ptr, e_conf.tok_clusterRawEnergy2	    , ele_vmaps, clusterRawEnergy2    );	
> > +  assignValue(ptr, e_conf.tok_clusterDPhiToSeed0    , ele_vmaps, clusterDPhiToSeed0   );
> > +  assignValue(ptr, e_conf.tok_clusterDPhiToSeed1    , ele_vmaps, clusterDPhiToSeed1   );
> > +  assignValue(ptr, e_conf.tok_clusterDPhiToSeed2    , ele_vmaps, clusterDPhiToSeed2   );
> > +  assignValue(ptr, e_conf.tok_clusterDEtaToSeed0    , ele_vmaps, clusterDEtaToSeed0   );
> > +  assignValue(ptr, e_conf.tok_clusterDEtaToSeed1    , ele_vmaps, clusterDEtaToSeed1   );
> > +  assignValue(ptr, e_conf.tok_clusterDEtaToSeed2    , ele_vmaps, clusterDEtaToSeed2   );
> > +  assignValue(ptr, e_conf.tok_iPhi, ele_int_vmaps, iPhi);
> > +  assignValue(ptr, e_conf.tok_iEta, ele_int_vmaps, iEta);
> > +  assignValue(ptr, e_conf.tok_cryPhi, ele_vmaps, cryPhi);
> > +  assignValue(ptr, e_conf.tok_cryEta, ele_vmaps, cryEta);
> > +
> > +  eval[8]  = eMax/sc->rawEnergy();
>
> enums maybe?
> We were discussing with Lindsey in another PR to have something like
> TMVAEvaluator for this kind of data manipulation.
> I guess a solution there is coming soon.
> @lgray <https://github.com/lgray>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10613/files#r36623288>.
>

slava77 commented on RecoEgamma/EgammaTools/plugins/EGExtraInfoModifierFromDB.cc:
```diff
@@ -0,0 +1,730 @@
+  float trkMomentumRelError = ele.trackMomentumError()/ele.trackMomentumAtVtx().R();
+  eval_ep[0] = tot_energy*mean;
+  eval_ep[1] = sigma/mean;
+  eval_ep[2] = ele.trackMomentumAtVtx().R(); 
```
please don't copy-paste the same function calls, define and use a temporary ``` ep = ele.trackMomentumAtVtx().R(); ```

matteosan1 committed:
small fix

slava77 commented:
@davidlange6 @smuzaffar @degano  
is it correct that if a binary file is introduced in one commit and then removed in another,
it will still show up in the pack file?
Do we have a cleanup step done occasionally to get rid of these cases?
the cms/git-cmssw-mirror/cmssw.git/objects/pack/ file is 0.8 GB.

cmsbuild commented:
Pull request #10613 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 137a6e56a8c52a1e38f52e1a6ff3bfb627d2efd0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7107/summary.html

slava77 commented:
From jenkins tests static checks
(from the link above to https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7107/llvm-analysis)
there are still issues in 
- EGExtraInfoModifierFromDB.cc has a memory leak https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7107/llvm-analysis/report-2e96bd.html#EndPath  This should be fixed.
- ElectronRegressionValueMapProducer.cc has a rather minor  issue https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7107/llvm-analysis/report-ab4b33.html#EndPath

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7107/summary.html

matteosan1 commented:
Fixed.

On 8/10/15 21:45, Slava Krutelyov wrote:
>
> From jenkins tests static checks
> (from the link above to 
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7107/llvm-analysis)
> there are still issues in
>
>   * EGExtraInfoModifierFromDB.cc has a memory leak
>     https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7107/llvm-analysis/report-2e96bd.html#EndPath
>     This should be fixed.
>   * ElectronRegressionValueMapProducer.cc has a rather minor issue
>     https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7107/llvm-analysis/report-ab4b33.html#EndPath
>
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/cms-sw/cmssw/pull/10613#issuecomment-129581412>.
>

+++++++++++
Matteo Sani
University of California, San Diego, US.
Tel: +41227671577
CERN Office: 40-3-A02

matteosan1 committed:
update to tag label and memory leak

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #10613 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

cmsbuild commented:
-1 
Tested at: 456da246120b962a60895b29310d3db1e8dd2526
 When I ran the RelVals I found an error in the following worklfows: 
 25202.0 step2
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

----- Begin Fatal Exception 11-Aug-2015 17:27:41 CEST-----------------------

An exception of category 'Configuration' occurred while

   [0] Processing run: 1 lumi: 1 event: 1

   [1] Running path 'digitisation_step'

   [2] Calling event method for module MixingModule/'mix'

Exception Message:

RootInputFileSequence::readOneRandom(): no input files specified for secondary input source.

----- End Fatal Exception -------------------------------------------------

</pre>
50202.0 step2
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50/step2_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50.log

----- Begin Fatal Exception 11-Aug-2015 17:28:25 CEST-----------------------

An exception of category 'Configuration' occurred while

   [0] Processing run: 1 lumi: 1 event: 1

   [1] Running path 'digitisation_step'

   [2] Calling event method for module MixingModule/'mix'

Exception Message:

RootInputFileSequence::readOneRandom(): no input files specified for secondary input source.

----- End Fatal Exception -------------------------------------------------

</pre>
4.22 step1
<pre>DAS Error</pre>
4.53 step1
<pre>DAS Error</pre>
140.53 step1
<pre>DAS Error</pre>
1000.0 step1
<pre>DAS Error</pre>
1001.0 step1
<pre>DAS Error</pre>
1003.0 step1
<pre>DAS Error</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10613/7146/summary.html

lgray commented:
@slava77 @matteosan1 taking over this PR too, I found .root files in the history and have removed them.

lgray commented:
@matteosan1 can you close this one? Thanks!
